pkg/sql/insert.go

```diff
@@ -746,35 +746,34 @@ func fillDefaults(
 		return defaultExprs[idx]
 	}
 
-	numColsOrig := len(ret.Tuples[0].Exprs)
+	numColsOrig := len(ret.Tuples[0])
 	for tIdx, tuple := range ret.Tuples {
-		if a, e := len(tuple.Exprs), numColsOrig; a != e {
+		if a, e := len(tuple), numColsOrig; a != e {
 			return nil, newValuesListLenErr(e, a)
 		}
 
 		tupleCopied := false
 		copyTuple := func() {
 			if !tupleCopied {
 				copyValues()
-				tuple = &tree.Tuple{Exprs: append([]tree.Expr(nil), tuple.Exprs...)}
-				ret.Tuples[tIdx] = tuple
+				ret.Tuples[tIdx] = append(tree.Exprs(nil), tuple...)
 				tupleCopied = true
 			}
 		}
 
-		for eIdx, val := range tuple.Exprs {
+		for eIdx, val := range tuple {
 			switch val.(type) {
 			case tree.DefaultVal:
 				copyTuple()
-				tuple.Exprs[eIdx] = defaultExpr(eIdx)
+				ret.Tuples[tIdx][eIdx] = defaultExpr(eIdx)
 			}
 		}
 
 		// The values for the row may be shorter than the number of columns being
 		// inserted into. Populate default expressions for those columns.
-		for i := len(tuple.Exprs); i < len(cols); i++ {
+		for i := len(tuple); i < len(cols); i++ {
 			copyTuple()
-			tuple.Exprs = append(tuple.Exprs, defaultExpr(len(tuple.Exprs)))
+			ret.Tuples[tIdx] = append(ret.Tuples[tIdx], defaultExpr(len(tuple)))
```

Comment:
@arjunravinarayan to follow-up on our conversation, this is the line where I introduced a bug.

The code should read instead:

```go
tuple = append(tuple, defaultExpr(len(tuple)))
ret.Tuples[tIdx] = tuple
```